lib/Backend/Commands.hs

```diff
@@ -0,0 +1,453 @@
+  entryOrDir <- getEntryOrDirThrow CPRPathNotFound oldPath
+
+  -- Build a list of operations to perform.
+  nowUtc <- embed getCurrentTime
+  operations <- fmap (setModifiedDate nowUtc) <$>
+    case entryOrDir of
+      Left entry -> copyEntry entry
+      Right dir -> copyDir dir
+
+  -- Validation checks.
+  checkDestinatationIsNotDirectory operations
+  when (not force) $ checkDestinatationIsNotEntry operations
+
+  -- Write new entries.
+  let newEntries = coNew <$> operations
+  forM_ newEntries writeSecret
+  pure $ CPRSuccess $ operations <&> \op -> (coOld op ^. E.path, coNew op ^. E.path)
+
+  where
+    copyEntry :: Entry -> Sem r [CopyOperation]
+    copyEntry entry =
+      -- If the `oldPath` points to an entry,
+      -- then `newPath` must also be a valid entry path.
+      case Path.pathAsEntryPath newPath of
+        Left _ -> throw CPRMissingEntryName
+        Right newEntryPath -> pure [CopyOperation entry (entry & E.path .~ newEntryPath)]
+
+    copyDir :: Directory -> Sem r [CopyOperation]
+    copyDir dir = do
+      forM (Dir.allEntries dir) \entry -> do
+        -- Get the entry's parent dir, remove the `oldPath` prefix from it,
+        -- and add the `newPath` prefix.
+        newEntry <- entry & E.path . Path.entryPathParentDir %%~ \parentDir ->
+          case Path.replacePathPrefix oldPath newPath parentDir of
+            Just newParentDir -> pure newParentDir
+            Nothing -> throw $ OtherError $ T.unlines
+              [ "Internal error:"
+              , "Expected path: '" <> pretty (entry ^. E.path) <> "'"
+              , "To have the prefix: '" <> pretty oldPath <> "'"
+              ]
+        pure $ CopyOperation entry newEntry
+
+    setModifiedDate :: UTCTime -> CopyOperation -> CopyOperation
+    setModifiedDate nowUtc (CopyOperation old new) =
+      CopyOperation old (new & dateModified .~ nowUtc)
+
+    -- | Checks that the paths we're trying to copy entries to aren't already occupied
+    -- by an existing directory.
+    checkDestinatationIsNotDirectory :: [CopyOperation] -> Sem r ()
+    checkDestinatationIsNotDirectory operations = do
+      invalidOperations <- flip filterM operations \(CopyOperation _ new) -> do
+        pathIsDirectory (new ^. E.path)
+      when (not (null invalidOperations)) do
+        invalidOperations
+          <&> (\(CopyOperation old new) -> (old ^. E.path, new ^. E.path) )
+          & throw . CPRDestinationIsDirectory
+
+    -- | Checks that the paths we're trying to copy entries to aren't already occupied
+    -- by an existing entry.
+    checkDestinatationIsNotEntry :: [CopyOperation] -> Sem r ()
+    checkDestinatationIsNotEntry operations = do
+      invalidOperations <- flip filterM operations \(CopyOperation _ new) -> do
+        pathIsEntry (new ^. E.path)
+      when (not (null invalidOperations)) do
+        invalidOperations
+          <&> (\(CopyOperation old new) -> (old ^. E.path, new ^. E.path) )
+          & throw . CPREntryAlreadyExists
```

Comment:
Looks like copy-paste. Maybe it would be better to create function `checkCopyOperations pred exception`?

Reply:
Done